File: lib/web/frontend/src/old_hash_redirector.ml
(*
 * This script will check old Laine-style hashes when the page has been loaded and redirects
 * to the correct address if such a hash is found.
 * This is done in JS because the hash is not sent to the server.
 *)

type hash_matcher =
  { old_hash : Js.Re.t;
    new_hash : string;
  }

let hash_re = [
  { old_hash = [%re "/^\\#\\!\\/(\\d{4})\\/(\\d\\d)\\/(\\d\\d)\\/(.*)$/"];
    new_hash = "/$1/$2/$3/$4";
  };
  { old_hash = [%re "/^\\#\\!\\/tag\\/([^\\/]+)$/"];
    new_hash = "/tag/$1";
  };
  { old_hash = [%re "/^\\#\\!\\/tag\\/([^\\/]+)(\\/\\d+)$/"];
    new_hash = "/tag/$1/p/$2";
  };
  { old_hash = [%re "/^\\#\\!\\/archives\\/(\\d{4})$/"];
    new_hash = "/archive/$1";
  };
  { old_hash = [%re "/^\\#\\!\\/archives\\/(\\d{4})\\/(\\d\\d)$/"];
    new_hash = "/archive/$1/$2";
  };
  { old_hash = [%re "/^\\#\\!\\/archives\\/(\\d{4})(\\/\\d+)$/"];
    new_hash = "/archive/$1/p/$2";
  };
  { old_hash = [%re "/\\#\\!\\/archives\\/(\\d{4})\\/(\\d\\d)(\\/\\d+)$/"];
    new_hash = "/archive/$1/$2/p/$3";
  };
  { old_hash = [%re "/^\\#\\!\\/(.*)$/"];
    new_hash = "/$1";
  }
]

let rec check_single_hash hash_list current_hash = match hash_list with
  | matcher :: tail ->
    let is_match = Js.Re.test current_hash matcher.old_hash in
      if is_match
        then Some(Js.String.replaceByRe matcher.old_hash matcher.new_hash current_hash)
        else check_single_hash tail current_hash
  | [] -> None

let run_hash_check hash =
  check_single_hash hash_re hash

File: lib/web/frontend/src/old-hash-redirector.ts
type ReList = ReadonlyArray<readonly [RegExp, string]>;

const HASH_RES: ReList = [
  [/^\#\!\/(\d{4})\/(\d\d)\/(\d\d)\/(.*)$/, "/$1/$2/$3/$4"],
  [/^\#\!\/tag\/([^\/]+)$/, "/tag/$1"],
  [/^\#\!\/tag\/([^\/]+)(\/\d+)$/, "/tag/$1/p/$2"],
  [/^\#\!\/archives\/(\d{4})$/, "/archive/$1"],
  [/^\#\!\/archives\/(\d{4})\/(\d\d)$/, "/archive/$1/$2"],
  [/^\#\!\/archives\/(\d{4})(\/\d+)$/, "/archive/$1/p/$2"],
  [/\#\!\/archives\/(\d{4})\/(\d\d)(\/\d+)$/, "/archive/$1/$2/p/$3"],
  [/^\#\!\/(.*)$/, "/$1"],
] as const;

export function oldHashRedirector() {
  const currentHash = window.location.hash;

  for (const [re, path] of HASH_RES) {
    if (re.test(path)) {
      window.location.href = currentHash.replace(re, path);
      return;
    }
  }
}
